source/crud/bulk-write.md

```diff
@@ -0,0 +1,899 @@
+exceeded. However, when an unacknowledged write concern is used, drivers MUST raise an error if one
+of the following limits is exceeded:
+
+- The size of a document to be inserted MUST NOT exceed `maxBsonObjectSize`. This applies to the
+  `document` field of an `InsertOneModel` and the `replacement` field of a `ReplaceOneModel`.
+- The size of an entry in the `ops` array MUST NOT exceed `maxBsonObjectSize + 16KiB`.
+- The size of the `bulkWrite` command document MUST NOT exceed `maxBsonObjectSize + 16KiB`.
+
+See [SERVER-10643](https://jira.mongodb.org/browse/SERVER-10643) for more details on these size
+limits.
+
+## Command Batching
+
+Drivers MUST accept an arbitrary number of operations as input to the `MongoClient.bulkWrite`
+method. Because the server imposes restrictions on the size of write operations, this means that a
+single call to `MongoClient.bulkWrite` may require multiple `bulkWrite` commands to be sent to the
+server. Drivers MUST split bulk writes into separate commands when the user's list of operations
+exceeds one or more of these maximums: `maxWriteBatchSize`, `maxBsonObjectSize` (for `OP_MSG`
+payload type 0), and `maxMessageSizeBytes` (for `OP_MSG` payload type 1). Each of these values can
+be retrieved from the selected server's `hello` command response. Drivers MUST merge results from
+multiple batches into a single `BulkWriteResult` or `BulkWriteException` to return from
+`MongoClient.bulkWrite`.
+
+When constructing the `nsInfo` array for a `bulkWrite` batch, drivers MUST only include the
+namespaces that are referenced in the `ops` array for that batch.
+
+### Number of Writes
+
+`maxWriteBatchSize` defines the total number of writes allowed in one command. Drivers MUST split a
+bulk write into multiple commands if the user provides more than `maxWriteBatchSize` operations in
+the argument for `models`.
+
+### Total Message Size
+
+#### Encrypted bulk writes
+
+When auto-encryption is enabled, drivers MUST NOT provide the `ops` and `nsInfo` fields as document
+sequences and MUST limit the size of each `bulkWrite` command according to the auto-encryption size
+limits defined in the
+[Client Side Encryption Specification](../client-side-encryption/client-side-encryption.rst).
+
+#### Unencrypted bulk writes
+
+When `ops` and `nsInfo` are provided as document sequences, drivers MUST ensure that the total size
+of the `OP_MSG` built for each `bulkWrite` command does not exceed `maxMessageSizeBytes`.
+
+The upper bound for the size of an `OP_MSG` includes opcode-related bytes (e.g. the `OP_MSG`
+header) and operation-agnostic command field bytes (e.g. `txnNumber`, `lsid`). Drivers MUST limit
+the combined size of the `bulkWrite`-specific command fields, `ops` document sequence, and `nsInfo`
+document sequence to `maxMessageSizeBytes - 1000` to account for this overhead. The following
+pseudocode demonstrates how to apply this limit in batch-splitting logic:
```

Comment:
Previous language for reference:

> Some drivers may perform batch-splitting prior to constructing the full `OP_MSG` to be sent to the server. In this case, drivers MAY use `maxMessageSizeBytes - 16,384` as the upper bound for the combined number of bytes in `ops` and `nsInfo`. 16KiB is subtracted as an approximate overhead allowance to accommodate for the bytes in the rest of the message.

The major difference here is that `bulkWrite` options are now excluded from the overhead allowance, which was problematic because _technically_ a `comment` or `let` document could be > 16,384 bytes and therefore cause the batch-splitting boundaries to be inaccurate.